app/services/catalog/copy_portfolio_item.rb

```diff
@@ -0,0 +1,70 @@
+module Catalog
+  class CopyPortfolioItem
+    attr_reader :new_portfolio_item
+
+    def initialize(params)
+      @name = params[:portfolio_item_name]
+      @portfolio_item = PortfolioItem.find(params[:portfolio_item_id])
+
+      begin
+        @to_portfolio = Portfolio.find(params[:portfolio_id] || @portfolio_item.portfolio_id)
+      rescue ActiveRecord::RecordNotFound
+        raise Catalog::InvalidParameter, "Portfolio specified not found"
+      end
+    end
+
+    def process
+      @new_portfolio_item = make_copy
+      @to_portfolio.portfolio_items << @new_portfolio_item
+
+      self
+    end
+
+    private
+
+    def make_copy
+      @portfolio_item.dup.tap do |new_portfolio_item|
+        new_portfolio_item.name = @name || new_name
+        new_portfolio_item.save
+      end
+    end
+
+    def new_name
+      if @portfolio_item.portfolio_id == @to_portfolio.id
+        create_copy_name
+      else
+        @portfolio_item.name
+      end
+    end
+
+    COPY_REGEX = '^Copy (\(\d\) )?of'.freeze
+
+    def create_copy_name
+      names = @to_portfolio.portfolio_items.collect(&:name)
```

Comment:
The usefulness of this regex check seems like it could possibly be built in the model associations. I'm up for suggestions otherwise, but it could pull some of the work out of this service class and put the collect_with_regex in the `portfolio.rb' model inline with the association. 
```
  has_many :portfolio_items, :dependent => :destroy do
    def collect_with_regex(regex, attr)
      list = Array.new.tap do |arr|
        self.each do |item|
          arr << item.send(attr).match(/#{regex} #{item.send(attr).sub(regex, "")}/)
        end
      end
      return list.compact
    end
  end
```

Reply:
:+1: implemented this with a few changes - cleaned up the code quite a bit!